pkg/controller/backup_configuration.go

```diff
@@ -0,0 +1,240 @@
+		if resource, err := c.dpLister.Deployments(namespace).Get(resource_name); err == nil {
+			key, err := cache.MetaNamespaceKeyFunc(resource)
+			if err == nil {
+				c.dpQueue.GetQueue().Add(key)
+			}
+			return err
+		}
+	case workload_api.KindDaemonSet:
+		if resource, err := c.dsLister.DaemonSets(namespace).Get(resource_name); err == nil {
+			key, err := cache.MetaNamespaceKeyFunc(resource)
+			if err == nil {
+				c.dsQueue.GetQueue().Add(key)
+			}
+			return err
+		}
+	case workload_api.KindStatefulSet:
+		if resource, err := c.ssLister.StatefulSets(namespace).Get(resource_name); err == nil {
+			key, err := cache.MetaNamespaceKeyFunc(resource)
+			if err == nil {
+				c.ssQueue.GetQueue().Add(key)
+			}
+		}
+	case workload_api.KindReplicationController:
+		if resource, err := c.rcLister.ReplicationControllers(namespace).Get(resource_name); err == nil {
+			key, err := cache.MetaNamespaceKeyFunc(resource)
+			if err == nil {
+				c.rcQueue.GetQueue().Add(key)
+			}
+			return err
+		}
+	case workload_api.KindReplicaSet:
+		if resource, err := c.rsLister.ReplicaSets(namespace).Get(resource_name); err == nil {
+			key, err := cache.MetaNamespaceKeyFunc(resource)
+			if err == nil {
+				c.rsQueue.GetQueue().Add(key)
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *StashController) EnsureCronJob(backupconfiguration *api_v1beta1.BackupConfiguration) error {
+	if backupconfiguration == nil {
+		return fmt.Errorf("BackupConfiguration is nil")
+	}
+	image := docker.Docker{
+		Registry: c.DockerRegistry,
+		Image:    docker.ImageStash,
+		Tag:      c.StashImageTag,
+	}
+
+	meta := metav1.ObjectMeta{
+		Name:      backupconfiguration.Name,
+		Namespace: backupconfiguration.Namespace,
+	}
+	ref, err := reference.GetReference(scheme.Scheme, backupconfiguration)
+	if err != nil {
+		return err
+	}
+	cronjob, _, err := batch_util.CreateOrPatchCronJob(c.kubeClient, meta, func(in *batch_v1beta1.CronJob) *batch_v1beta1.CronJob {
+		//set backup-configuration as cron-job owner
+		core_util.EnsureOwnerReference(&in.ObjectMeta, ref)
+
+		in.Spec.Schedule = backupconfiguration.Spec.Schedule
+		if in.Spec.JobTemplate.Labels == nil {
+			in.Spec.JobTemplate.Labels = map[string]string{}
+		}
+		in.Spec.JobTemplate.Labels[util.LabelApp] = util.AppLabelStash
+		in.Spec.JobTemplate.Spec.Template.Spec.Containers = core_util.UpsertContainer(
+			in.Spec.JobTemplate.Spec.Template.Spec.Containers,
+			core.Container{
+				Name:            backupconfiguration.Name,
+				ImagePullPolicy: core.PullIfNotPresent,
+				Image:           image.ToContainerImage(),
+				Args: []string{
+					"backup-session",
+					fmt.Sprintf("--backupsession.name=%s", backupconfiguration.Name),
+					fmt.Sprintf("--backupsession.namespace=%s", backupconfiguration.Namespace),
+				},
+			})
+		in.Spec.JobTemplate.Spec.Template.Spec.RestartPolicy = core.RestartPolicyNever
+		if c.EnableRBAC {
+			in.Spec.JobTemplate.Spec.Template.Spec.ServiceAccountName = in.Name
```

Comment:
We should ensure service account first then we can use the service account. `EnsureCronJobRolebinding` before creating cronjob. Set `BackupConfiguration`  as owner of the RoleBinding and service account.

Reply:
resolved